src/java.base/share/classes/com/sun/crypto/provider/HkdfKeyDerivation.java

```diff
@@ -0,0 +1,399 @@
+
+                    bb = Arrays.copyOf(bb, bb.length + workItemBytes.length);
+                    System.arraycopy(workItemBytes, 0, bb,
+                                     bb.length - workItemBytes.length,
+                                     workItemBytes.length);
+                }
+                return new SecretKeySpec(bb, "Generic");
+            }
+        } else if (keys.isEmpty()) {
+            return null;
+        } else {
+            throw new InvalidKeyException(
+                "List of keys could not be consolidated");
+        }
+    }
+
+    /**
+     * Perform the HMAC-Extract operation.
+     *
+     * @param inputKey
+     *     the input keying material used for the HKDF-Extract operation.
+     * @param salt
+     *     the salt value used for HKDF-Extract.  If no salt is to be used a
+     *     {@code null} value should be provided.
+     *
+     * @return a byte array containing the pseudorandom key (PRK)
+     *
+     * @throws InvalidKeyException
+     *     if an invalid salt was provided through the
+     *     {@code HkdfParameterSpec}
+     */
+    protected byte[] hkdfExtract(SecretKey inputKey, byte[] salt)
+        throws InvalidKeyException, NoSuchAlgorithmException {
+
+        if (salt == null) {
+            salt = new byte[hmacLen];
+        }
+        Mac hmacObj = Mac.getInstance(hmacAlgName);
+        hmacObj.init(new SecretKeySpec(salt, "HKDF-Salt"));
+
+        if (inputKey == null) {
+            return hmacObj.doFinal();
+        } else {
+            return hmacObj.doFinal(inputKey.getEncoded());
+        }
+    }
+
+    /**
+     * Perform the HMAC-Expand operation.  At the end of the operation, the
+     * keyStream instance variable will contain the complete KDF output based on
+     * the input values and desired length.
+     *
+     * @param prk
+     *     the pseudorandom key used for HKDF-Expand
+     * @param info
+     *     optional context and application specific information or {@code null}
+     *     if no info data is provided.
+     * @param outLen
+     *     the length in bytes of the required output
+     *
+     * @return a byte array containing the complete KDF output.  This will be at
+     *     least as long as the requested length in the {@code outLen}
+     *     parameter, but will be rounded up to the nearest multiple of the HMAC
+     *     output length.
+     *
+     * @throws InvalidKeyException
+     *     if an invalid key was provided through the {@code HkdfParameterSpec}
+     *     or derived during the generation of the PRK.
+     */
+    protected byte[] hkdfExpand(SecretKey prk, byte[] info, int outLen)
+        throws InvalidKeyException, NoSuchAlgorithmException {
+        byte[] kdfOutput;
+        Mac hmacObj = Mac.getInstance(hmacAlgName);
+
+        // Calculate the number of rounds of HMAC that are needed to
+        // meet the requested data.  Then set up the buffers we will need.
+        hmacObj.init(prk);
```

Comment:
RFC5869 sec 2.3 states that "PRK - a pseudorandom key of at least HashLen octets". Shouldn't we check it before passing to to `hmacObj`?

Reply:
Well, if a non-extractrable hardware key is passed to us, it'd lead to NPE with the new check. Instead of calling getEncoded() on it and access  length directly, maybe relying on `CipherCore.getKeyBytes(prk)` which will throw InvalidKeyException?

Reply:
Given that this comment is marked resolved, I am going to copy the above comment to the new file and start a new conversation so it's more visible.